File: nara_service/backend/app/routers/didyouknow.py
"""
Did You Know Router - 흥미로운 공공데이터 사실 API

Why: 공공데이터 API 문서에서 LLM으로 생성한 흥미로운 사실을 제공합니다.
"""

from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional
from app.services.didyouknow_service import DidYouKnowService, FactCategory
from app.domain.common.dependencies import get_didyouknow_service

router = APIRouter(prefix="/didyouknow", tags=["Did You Know"])


# ==================== 모델 정의 ====================

class FactResponse(BaseModel):
    """단일 사실 응답"""
    id: str = Field(..., description="고유 ID (UUID)")
    category: str = Field(..., description="카테고리")
    content: str = Field(..., description="사실 내용 (한 문장)")
    source_doc_id: Optional[str] = Field(None, description="원천 문서 ID")
    created_at: str = Field(..., description="생성 시간 (ISO 8601)")
    metadata: Dict[str, Any] = Field(default_factory=dict, description="추가 메타데이터")


class FactsListResponse(BaseModel):
    """사실 목록 응답"""
    total: int = Field(..., description="총 개수")
    facts: List[FactResponse] = Field(..., description="사실 목록")


class LLMParams(BaseModel):
    """LLM 생성 파라미터"""
    temperature: float = Field(0.8, ge=0.0, le=1.0, description="Temperature (0.0-1.0)")
    top_p: float = Field(0.9, ge=0.0, le=1.0, description="Top P (0.0-1.0)")
    max_tokens: int = Field(150, ge=50, le=500, description="최대 토큰 수")


class CategoryCounts(BaseModel):
    """카테고리별 생성 개수"""
    api_introduction: int = Field(34, ge=0, le=100, description="API 소개 개수")
    provider_introduction: int = Field(34, ge=0, le=100, description="제공 기관 소개 개수")
    usage_tip: int = Field(34, ge=0, le=100, description="활용 팁 개수")


class GenerateRequest(BaseModel):
    """콘텐츠 생성 요청 (관리자용)"""
    counts: CategoryCounts = Field(..., description="카테고리별 생성 개수")
    llm_params: LLMParams = Field(default_factory=LLMParams, description="LLM 생성 파라미터")


# ==================== 엔드포인트 ====================

@router.get("/facts", response_model=FactsListResponse)
async def get_all_facts(
    category: Optional[str] = None,
    service: DidYouKnowService = Depends(get_didyouknow_service)
) -> FactsListResponse:
    """
    저장된 모든 사실 조회

    Why: 프론트엔드에서 슬라이드쇼에 표시할 사실 목록을 가져옵니다.
         캐시된 facts.json에서 로드하므로 매우 빠릅니다 (1ms 이내).

    Query Parameters:
        category: 카테고리 필터 (선택)
            - api_introduction: 특정 API 소개
            - provider_introduction: 제공 기관 소개
            - usage_tip: 데이터 활용 팁

    Returns:
        사실 목록 (total, facts)

    Raises:
        HTTPException: 로드 실패 시 500 에러
    """
    try:
        facts = service.load_facts()

        if category:
            facts = [f for f in facts if f.get('category') == category]

        return FactsListResponse(
            total=len(facts),
            facts=facts
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to load facts: {str(e)}")


@router.get("/random", response_model=FactResponse)
async def get_random_fact(
    category: Optional[str] = None,
    service: DidYouKnowService = Depends(get_didyouknow_service)
) -> FactResponse:
    """
    랜덤 사실 1개 반환

    Why: 매번 다른 사실을 보여주거나 API 테스트용으로 사용합니다.

    Query Parameters:
        category: 카테고리 필터 (선택)

    Returns:
        랜덤 사실 1개

    Raises:
        HTTPException: 사실이 없으면 404, 로드 실패 시 500
    """
    try:
        fact = service.get_random_fact(category)

        if not fact:
            raise HTTPException(
                status_code=404,
                detail="No facts available" + (f" for category: {category}" if category else "")
            )

        return fact
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get random fact: {str(e)}")


@router.post("/generate")
async def generate_facts(
    request: GenerateRequest,
    service: DidYouKnowService = Depends(get_didyouknow_service)
) -> Dict[str, Any]:
    """
    새로운 콘텐츠 생성 (관리자용)

    Why: 캐시된 사실이 없거나 새로운 콘텐츠를 생성하고 싶을 때 사용합니다.
         LLM 호출이 필요하므로 5-10분 소요될 수 있습니다.
         기존 facts.json을 삭제하고 새로 생성합니다.

    Body:
        category: 특정 카테고리만 생성 (선택, null이면 전체)
        count_per_category: 카테고리당 생성 개수 (기본 34개, 3개 카테고리 = 총 102개)

    Returns:
        생성 결과 (generated_count, total_count)

    Technical Flow:
        1단계: RAG 시스템에서 원천 문서 선택
        2단계: 카테고리별 LLM 프롬프트 생성
        3단계: Ollama gemma3:1b로 콘텐츠 생성
        4단계: 기존 파일 삭제 후 새로 생성된 사실만 저장

    Raises:
        HTTPException: 생성 실패 시 500 에러
    """
    try:
        # 카테고리별 개수 설정
        counts = {
            FactCategory.API_INTRODUCTION: request.counts.api_introduction,
            FactCategory.PROVIDER_INTRO: request.counts.provider_introduction,
            FactCategory.USAGE_TIP: request.counts.usage_tip,
        }

        # LLM 파라미터 추출
        llm_params = {
            "temperature": request.llm_params.temperature,
            "top_p": request.llm_params.top_p,
            "max_tokens": request.llm_params.max_tokens,
        }

        # 새 사실 생성 (LLM 파라미터 전달)
        new_facts = service.generate_batch(counts, llm_params)

        # 기존 파일 삭제하고 새로 생성된 사실만 저장
        service.save_facts(new_facts)

        return {
            "status": "success",
            "generated_count": len(new_facts),
            "total_count": len(new_facts)
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to generate facts: {str(e)}")


@router.get("/stats")
async def get_stats(
    service: DidYouKnowService = Depends(get_didyouknow_service)
) -> Dict[str, Any]:
    """
    통계 정보 반환

    Why: 관리자가 현재 캐시된 사실의 현황을 파악하기 위해 사용합니다.

    Returns:
        통계 정보 (total, by_category)

    Raises:
        HTTPException: 로드 실패 시 500 에러
    """
    try:
        facts = service.load_facts()

        stats = {
            "total": len(facts),
            "by_category": {}
        }

        for cat in FactCategory:
            stats["by_category"][cat.value] = len([f for f in facts if f.get('category') == cat.value])

        return stats

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get stats: {str(e)}")

File: nara_service/frontend/src/lib/api/didyouknow.ts
/**
 * Did You Know API - 흥미로운 공공데이터 사실
 *
 * Why: Result Pattern을 적용하여 타입 안전한 에러 처리를 제공합니다.
 */
import { BASE_URL, HEADERS, fetchWithResult } from './client';
import type { Result } from '@/lib/types/result';

/**
 * 사실 인터페이스
 */
export interface Fact {
  id: string;
  category: string;
  content: string;
  source_doc_id?: string;
  created_at: string;
  metadata?: Record<string, any>;
}

/**
 * 사실 목록 응답
 */
export interface FactsListResponse {
  total: number;
  facts: Fact[];
}

/**
 * 통계 응답
 */
export interface StatsResponse {
  total: number;
  by_category: Record<string, number>;
}

/**
 * 모든 사실 조회
 *
 * Why: Result Pattern을 사용하여 타입 안전한 에러 처리를 제공합니다.
 *
 * @param category 카테고리 필터 (선택)
 * @returns Result<사실 목록>
 */
export const getAllFacts = async (
  category?: string
): Promise<Result<FactsListResponse>> => {
  const url = category
    ? `${BASE_URL}/didyouknow/facts?category=${category}`
    : `${BASE_URL}/didyouknow/facts`;

  return await fetchWithResult<FactsListResponse>(url, {
    method: 'GET',
    headers: HEADERS,
  });
};

/**
 * 랜덤 사실 1개 조회
 *
 * Why: Result Pattern을 사용하여 타입 안전한 에러 처리를 제공합니다.
 *
 * @param category 카테고리 필터 (선택)
 * @returns Result<단일 사실>
 */
export const getRandomFact = async (
  category?: string
): Promise<Result<Fact>> => {
  const url = category
    ? `${BASE_URL}/didyouknow/random?category=${category}`
    : `${BASE_URL}/didyouknow/random`;

  return await fetchWithResult<Fact>(url, {
    method: 'GET',
    headers: HEADERS,
  });
};

/**
 * 통계 조회
 *
 * Why: Result Pattern을 사용하여 타입 안전한 에러 처리를 제공합니다.
 *
 * @returns Result<통계 정보>
 */
export const getStats = async (): Promise<Result<StatsResponse>> => {
  return await fetchWithResult<StatsResponse>(`${BASE_URL}/didyouknow/stats`, {
    method: 'GET',
    headers: HEADERS,
  });
};

/**
 * 생성 결과 응답
 */
export interface GenerateResponse {
  status: string;
  generated_count: number;
  total_count: number;
}

/**
 * 생성 설정 인터페이스
 */
export interface GenerateConfig {
  counts: {
    api_introduction: number;
    provider_introduction: number;
    usage_tip: number;
  };
  llm_params: {
    temperature: number;
    top_p: number;
    max_tokens: number;
  };
}

/**
 * 신문기사 정보 인터페이스
 */
export interface ArticleInfo {
  index: number;
  title: string;
  url: string;
  count: number;
  article_preview: string;
}

/**
 * 관련 API 응답 인터페이스
 */
export interface RelatedAPI {
  id: string;
  title: string;
  description: string;
  provider: string;
  keywords: string[];
  url: string;
  type: string;
  score: number;
  rank_method?: string; // 하이브리드 RAG용 (llm_reranked, keyword_fallback)
}

/**
 * 신문기사와 관련 API 인터페이스
 */
export interface ArticleWithAPIs {
  index: number;
  title: string;
  url: string;
  count: number;
  article_preview: string;
  related_apis: RelatedAPI[];
  match_method?: string; // hybrid_rag, keyword_only
}

/**
 * 새 콘텐츠 생성 (관리자용)
 *
 * Why: Result Pattern을 사용하여 타입 안전한 에러 처리를 제공합니다.
 *      생성에 5-10분이 걸릴 수 있으므로 timeout을 15분으로 설정합니다.
 *
 * @param config 생성 설정 (카테고리별 개수, LLM 파라미터)
 * @returns Result<생성 결과>
 */
export const generateFacts = async (
  config: GenerateConfig
): Promise<Result<GenerateResponse>> => {
  // 15분 timeout (생성 시간이 길 수 있음)
  const LONG_TIMEOUT = 15 * 60 * 1000; // 900000ms

  return await fetchWithResult<GenerateResponse>(
    `/api/backend/didyouknow/generate`,
    {
      method: 'POST',
      headers: HEADERS,
      body: JSON.stringify(config),
    },
    LONG_TIMEOUT
  );
};

/**
 * 모든 신문기사 목록 조회
 *
 * @returns Result<신문기사 목록>
 */
export const getArticles = async (): Promise<Result<ArticleInfo[]>> => {
  return await fetchWithResult<ArticleInfo[]>(`${BASE_URL}/didyouknow/articles`, {
    method: 'GET',
    headers: HEADERS,
  });
};

/**
 * 신문기사 내용 기반으로 관련 API 문서 찾기
 *
 * @param articleIndex 기사 인덱스 (1부터 시작)
 * @param articleText 직접 제공된 기사 텍스트 (선택)
 * @param topK 반환할 문서 개수 (기본값: 20)
 * @returns Result<관련 API 목록>
 */
export const getRelatedAPIs = async (
  articleIndex?: number,
  articleText?: string,
  topK: number = 20
): Promise<Result<RelatedAPI[]>> => {
  const params = new URLSearchParams();
  if (articleIndex) params.append('article_index', articleIndex.toString());
  if (articleText) params.append('article_text', articleText);
  params.append('top_k', topK.toString());

  const url = `${BASE_URL}/didyouknow/related-apis?${params.toString()}`;

  return await fetchWithResult<RelatedAPI[]>(url, {
    method: 'GET',
    headers: HEADERS,
  });
};

/**
 * 모든 신문기사와 관련 API 목록 조회 (키워드 기반, 미리 계산됨)
 *
 * @param topK 각 기사당 반환할 API 개수 (기본값: 20)
 * @param minScore 최소 유사도 점수 (기본값: 25.0, 키워드 유사도 0–100 스케일)
 * @returns Result<신문기사와 관련 API 목록>
 */
export const getArticlesWithAPIs = async (
  topK: number = 20,
  minScore: number = 25.0
): Promise<Result<ArticleWithAPIs[]>> => {
  const params = new URLSearchParams();
  params.append('top_k', topK.toString());
  params.append('min_score', minScore.toString());

  const url = `${BASE_URL}/didyouknow/articles-with-apis?${params.toString()}`;

  return await fetchWithResult<ArticleWithAPIs[]>(url, {
    method: 'GET',
    headers: HEADERS,
  });
};

/**
 * 하이브리드 RAG로 매칭된 신문기사와 관련 API 목록 조회
 *
 * 2단계 접근 방식:
 * 1. 키워드 필터링 (빠름, 100k → 30 후보)
 * 2. LLM 의미적 재랭킹 (gemma3:4b, 30 → 5 최종)
 *
 * @param useCache 캐시된 결과 사용 여부 (기본값: true)
 * @param forceRegenerate 캐시 무시하고 강제 재생성 (기본값: false)
 * @returns Result<신문기사와 관련 API 목록>
 */
export const getArticlesWithAPIsHybrid = async (
  useCache: boolean = true,
  forceRegenerate: boolean = false
): Promise<Result<ArticleWithAPIs[]>> => {
  const params = new URLSearchParams();
  params.append('use_cache', useCache.toString());
  params.append('force_regenerate', forceRegenerate.toString());

  const url = `${BASE_URL}/didyouknow/articles-with-apis-hybrid?${params.toString()}`;

  // 재생성 시 최대 2분 대기 (20기사 × 4초)
  const timeout = forceRegenerate ? 120000 : 30000;

  return await fetchWithResult<ArticleWithAPIs[]>(url, {
    method: 'GET',
    headers: HEADERS,
  }, timeout);
};

/**
 * 재생성 응답 인터페이스
 */
export interface RegenerateResponse {
  status: string;
  generated_at: string;
  total_articles: number;
  total_matches: number;
  avg_matches_per_article: number;
  method: string;
  message: string;
}

/**
 * 신문기사-API 매칭 수동 재생성
 *
 * 사용 사례:
 * - 새 media_links_*.json 파일 업로드
 * - index.json 업데이트 (새 API 추가)
 * - 매칭 알고리즘 파라미터 변경 후 테스트
 *
 * @returns Result<재생성 결과>
 */
export const regenerateArticleMatches = async (): Promise<Result<RegenerateResponse>> => {
  // 재생성에 최대 2분 소요
  const REGENERATE_TIMEOUT = 120000;

  return await fetchWithResult<RegenerateResponse>(
    `${BASE_URL}/didyouknow/regenerate-article-matches`,
    {
      method: 'POST',
      headers: HEADERS,
    },
    REGENERATE_TIMEOUT
  );
};
